File: simple_poker_hands_rust/src/poker.rs
use crate::cards::Card;

type Kicker = Card;

#[derive(Debug, PartialEq)]
pub enum HandEnum {
    HighCardEnum(HighCard),
    PairEnum(Pair),
    TwoPairEnum(TwoPair),
    ThreeOfAKindEnum(ThreeOfAKind),
    StraightEnum(Straight),
    FlushEnum(Flush),
    FullHouseEnum(FullHouse),
    FourOfAKindEnum(FourOfAKind),
    StraightFlushEnum(StraightFlush),
}

#[derive(Debug, PartialEq)]
pub struct HighCard {
    pub card: Card,
}
#[derive(Debug, PartialEq)]
pub struct Pair {
    pub cards: (Card, Card),
    pub kicker: Kicker,
}
#[derive(Debug, PartialEq)]
pub struct TwoPair {
    pub first: (Card, Card),
    pub kicker: Kicker,
    pub second: (Card, Card),
}
#[derive(Debug, PartialEq)]
pub struct ThreeOfAKind {
    pub cards: (Card, Card, Card),
    pub kicker: Kicker,
}
#[derive(Debug, PartialEq)]
pub struct Straight {
    pub cards: (Card, Card, Card, Card, Card),
}
#[derive(Debug, PartialEq)]
pub struct Flush {
    pub cards: (Card, Card, Card, Card, Card),
}
#[derive(Debug, PartialEq)]
pub struct FullHouse {
    pub first: (Card, Card, Card),
    pub second: (Card, Card),
}
#[derive(Debug, PartialEq)]
pub struct FourOfAKind {
    pub cards: (Card, Card, Card, Card),
    pub kicker: Kicker,
}
#[derive(Debug, PartialEq)]
pub struct StraightFlush {
    pub cards: (Card, Card, Card, Card, Card),
}

File: simple_poker_hands_kotlin/src/main/kotlin/Poker.kt
import kotlin.collections.map

typealias Kicker = Card

sealed class PokerHand {
    data class HighCard(
        val card: Card,
    ) : PokerHand()

    data class OnePair(
        val cards: Pair<Card, Card>,
        val kicker: Kicker,
    ) : PokerHand()

    data class TwoPair(
        val first: Pair<Card, Card>,
        val second: Pair<Card, Card>,
        val kicker: Kicker,
    ) : PokerHand()

    data class ThreeOfAKind(
        val cards: Triple<Card, Card, Card>,
        val kicker: Kicker,
    ) : PokerHand()

    data class Straight(
        val cards: Quintuple<Card, Card, Card, Card, Card>,
    ) : PokerHand() {
        init {
            val cardsList = cards.toList().sortedBy { it.strength.ordinal }
            require(cardsList.zipWithNext().all { (a, b) -> b.strength.ordinal == a.strength.ordinal + 1 })
        }
    }

    data class Flush(
        val cards: Quintuple<Card, Card, Card, Card, Card>,
    ) : PokerHand() {
        init {
            require(
                cards
                    .toList()
                    .map { it.suit }
                    .distinct()
                    .size == 1,
            )
        }
    }

    data class FullHouse(
        val first: Triple<Card, Card, Card>,
        val second: Pair<Card, Card>,
    ) : PokerHand()

    data class FourOfAKind(
        val cards: Quadruple<Card, Card, Card, Card>,
        val kicker: Kicker,
    ) : PokerHand()

    data class StraightFlush(
        val cards: Quintuple<Card, Card, Card, Card, Card>,
    ) : PokerHand()
}
